src/librustc_typeck/check/expr.rs

```diff
@@ -147,8 +148,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         debug!(">> type-checking: expr={:?} expected={:?}",
                expr, expected);
 
+        // If when desugaring the try block we ok-wrapped an expression that diverges
+        // (e.g. `try { return }`) then technically the ok-wrapping expression is unreachable.
+        // But since it is autogenerated code the resulting warning is confusing for the user
+        // so we want avoid generating it.
+        // Ditto for the autogenerated `Try::from_ok(())` at the end of e.g. `try { return; }`.
+        let (is_try_block_ok_wrapped_expr, is_try_block_generated_expr) = match expr.node {
+            ExprKind::Call(_, ref args) if expr.span.is_desugaring(DesugaringKind::TryBlock) => {
```

Comment:
This code is rather brittle. I wonder if it would be better to just ok-wrap the whole try-block rather than the final expression.

Reply:
Well I tried that (desugaring `try {...}` to `{ Try::from_ok({...}) }`) and the resulting code is indeed simpler (didn't push yet). The downside is that type errors now refer to the whole block and not to the tail expression (e.g. [here](https://github.com/rust-lang/rust/blob/9b9d2aff8de4d499b4ba7ca406e000f8d3754ea7/src/test/ui/try-block/try-block-type-error.rs#L10)). Thoughts?